Assets/BossRoom/Scripts/Client/UI/IPUIMediator.cs

```diff
@@ -0,0 +1,191 @@
+            NameGenerationData nameGenerationData,
+            GameNetPortal gameNetPortal,
+            ClientGameNetPortal clientGameNetPortal,
+            LobbyUIMediator lobbyUIMediator
+        )
+        {
+            m_NameGenerationData = nameGenerationData;
+            m_GameNetPortal = gameNetPortal;
+            m_ClientNetPortal = clientGameNetPortal;
+
+            RegenerateName();
+
+            m_ClientNetPortal.ConnectFinished += OnConnectFinished;
+        }
+
+        void OnDestroy()
+        {
+            if (m_ClientNetPortal != null)
+            {
+                m_ClientNetPortal.ConnectFinished -= OnConnectFinished;
+            }
+        }
+
+        void Awake()
+        {
+            Hide();
+        }
+
+        void Start()
+        {
+            // show create IP as default
+            ToggleCreateIPUI();
+        }
+
+        void Update()
+        {
+            if (Input.touchCount == k_NbTouchesToOpenWindow && AnyTouchDown() ||
+                m_OpenIPWindowKeyCode != KeyCode.None && Input.GetKeyDown(m_OpenIPWindowKeyCode))
+            {
+                if (m_CanvasGroup.interactable)
+                {
+                    Hide();
+                }
+                else
+                {
+                    Show();
+                }
+            }
+        }
+
+        static bool AnyTouchDown()
+        {
+            foreach (var touch in Input.touches)
+            {
+                if (touch.phase == TouchPhase.Began)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        public void HostIPRequest(string ip, string port)
+        {
+            int.TryParse(port, out var portNum);
+            if (portNum <= 0)
+            {
+                portNum = k_DefaultPort;
+            }
+
+            ip = string.IsNullOrEmpty(ip) ? k_DefaultIP : ip;
+
+            m_GameNetPortal.PlayerName = m_PlayerNameLabel.text;
+
+            m_GameNetPortal.StartHost(ip, portNum);
+
+            m_SignInSpinner.SetActive(true);
+        }
+
+        public void JoinWithIP(string ip, string port)
+        {
+            int.TryParse(port, out var portNum);
+            if (portNum <= 0)
+            {
+                portNum = k_DefaultPort;
+            }
+
+            ip = string.IsNullOrEmpty(ip) ? k_DefaultIP : ip;
+
+            m_GameNetPortal.PlayerName = m_PlayerNameLabel.text;
+
+            m_ClientNetPortal.StartClient(ip, portNum);
+
+            m_SignInSpinner.SetActive(true);
+        }
+
+        public void RegenerateName()
+        {
+            m_PlayerNameLabel.text = m_NameGenerationData.GenerateName();
+        }
+
+        /// <summary>
+        /// Callback when the server sends us back a connection finished event.
+        /// </summary>
+        /// <param name="status"></param>
+        void OnConnectFinished(ConnectStatus status)
+        {
+            // TODO: publish ConnectStatus message
```

Comment:
I'm guessing this is waiting on LP's popup? which is waiting on your approval? :P

Reply:
Correct! I suppose that PR is a dependency for this one.

Reply:
This shouldn't be necessary after the other pr, the connectFinished callback is removed and the connect status is published inside ClientGameNetPortal instead.

Reply:
popup merged